### ISSUE TITLE
Kernel keep dying when running lightGBM

### ISSUE THREAD

**Issue by TengGao** [flennerhag/mlens#47]
I ran the sample kernel script from kaggle: https://www.kaggle.com/flennerhag/ml-ensemble-scikit-learn-style-ensemble-learning

The script will run without issue. But, when I add a lightGBM to the base learner, the script will run for hours without finishing the calculation. I changed the nthread=1 (from -1), and set the Evaluator to backend='threading', njobs=1, and remove the xgboost from the base learner, the kernel will die whenever running the fit method. 

Here is the parameter for the lightgbm:
lgb = LGBMRegressor(objective='regression', nthread=1,seed=SEED)
'lgb':
                   {'learning_rate': uniform(0.02,0.04),
                    'num_leaves': randint(50, 60),
                    'n_estimators': randint(150,200),
                    'min_child_weight': randint(30,60)}

setup for the Evaluator:

evl = Evaluator(scorer,
                cv=2,
                random_state=SEED,
                verbose=5,
                backend='threading',
                n_jobs=1
               )

Please let me know if I need to provide any more info.
best
Mike

**Comment by flennerhag** [comment]
Hi, 

Could you specify:
- are you running the notebook on Kaggle's docker image or on your own machine? If the latter, what's the python version and the os?
- which ``fit`` call is it that stalls (i.e. which code block)? The first evaluator or the second?
- where exactly do you put the lgbm model? I.e. as a model to be evaluated, or as model in the ``EnsembleTransformer``? If the latter, please try setting ``backend='threading'`` on the EnsembleTransformer as well, and possibly ``n_jobs=1``.
- finally, could you check which version of ML-Ensemble you are running?
      ``import mlens; print(mlens.__version__)``

Thanks

**Comment by TengGao** [comment]
Hi:

1) I am running the program in my own machine, using python 3.6 with Ubuntu
16.04
2) The fit method is referring to the second evaluator, on code block 17
3) The lgbm is used as the base model only
4) The mlens  is version 0.1.5.2

Thanks
Mike

**Comment by flennerhag** [comment]
I'm having trouble replicating your problem. Could you try the following and see if it works?

First, put the following in a script:

```
import numpy as np
from lightgbm import LGBMRegressor
from mlens.model_selection import Evaluator
from mlens.metrics import rmse, make_scorer
from scipy.stats import uniform, randint

if __name__ == '__main__':
    X = np.arange(1000).reshape(100, 10)
    y = np.random.rand(100)

    scorer = make_scorer(rmse)

    evl = Evaluator(scorer, n_jobs=-1, backend='multiprocessing', verbose=100)
    evl.fit(X, y, [('model', LGBMRegressor(nthread=-1))],
             {'model':
                      {'learning_rate': uniform(0.02,0.04),
                       'num_leaves': randint(50, 60),
                      'n_estimators': randint(150,200),
                      'min_child_weight': randint(30,60)}
             }
          )
```

Then run the script from the command line. The script should complete in a few seconds. Does this work for you?

If it does, could you try pasting it into a jupyter notebook and run the notebook. Does that work?

**Comment by TengGao** [comment]
I had run your code both in command line and in a jupyter notebook, and it
both ran fine.

Here is the jupyter notebook of the code I used that ran into problem:

https://github.com/TengGao/ML_Ensemble/blob/master/ML_Ensemble.ipynb

It is the second fit method for the meta learner. The notebook will
normally either stall, or take a long time to run. It also only uses 12% of
the CPU.

Thank you for your quick response!
Mike

**Comment by TengGao** [comment]
This is exciting! I am not exactly sure how to run the current master
branch. Should I remove the current version from my machine and follow the
instruction on downloading the latest version?

I will give it a try.

**Comment by flennerhag** [comment]
Thanks! 

I recommend setting up a virtual environment. If you are on anaconda, use conda create
```
conda create --name test-env numpy scipy scikit-learn
```
Otherwise, you can use ``virtualenv``
```
virtualenv --system-site-packages -p python3 test-env
```

Activate the virtual environment with

```
source activate test-env
```

To install the bleeding edge:
```
git clone https://github.com/flennerhag/mlens.git; cd mlens;
git checkout master
python install setup.py
```
And you're good to go. 

Once done, deactivate environment using ``source deactivate`` and then remove the virtual environment: 
```
conda remove --name flowers --all
```
on conda, and on ``virtualenv`` recursively remove the directory (``rm -rf test-env``).

**Comment by TengGao** [comment]
Thank you! I will give it a try later today.

**Comment by TengGao** [comment]
I had download and tried the latest version on a AWS EC2 instance with
Ubuntu 16.04 (the same set up as my local machine). The program will run
without problem.

The detail log of the program that I ran can be found here:
https://github.com/TengGao/ML_Ensemble/blob/master/ML_Ensemble.ipynb

However, when I add the parameter 'colsample_bytree': randint(1,10)' to the
grid_search, it will throw an error. Other than that, it runs fine.

Thank you!
Mike

**Comment by flennerhag** [comment]
Great, I'll close this issue then. 

(The ``colsample_bytree`` is due to sampling more (or less) features that what is available. Set a smaller range e.g. 1-5)

Thanks for the assist!

**Comment by waylongo** [comment]
try n_jobs != -1
